pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -73,14 +203,117 @@ func (a *saOAuthClientAdapter) GetClient(ctx kapi.Context, name string) (*oautha
 		// TODO update this to allow https redirection to any
 		// 1. service IP (useless in general)
 		// 2. service DNS (useless in general)
-		// 3. route DNS (useful)
-		// 4. loopback? (useful, but maybe a bit weird)
-		RedirectURIs: redirectURIs,
+		// 3. loopback? (useful, but maybe a bit weird)
+		RedirectURIs: sets.NewString(redirectURIs...).List(),
 		GrantMethod:  a.grantMethod,
 	}
 	return saClient, nil
 }
 
+func parseModelsMap(annotations map[string]string) map[string]model {
+	models := map[string]model{}
+	for key, value := range annotations {
+		if prefix, name, ok := parseModelPrefixName(key); ok {
+			m := models[name]
+			switch prefix {
+			case OAuthRedirectModelAnnotationURISchemePrefix:
+				m.scheme = value
+			case OAuthRedirectModelAnnotationURIPortPrefix:
+				m.port = value
+			case OAuthRedirectModelAnnotationURIPathPrefix:
+				m.path = value
+			case OAuthRedirectModelAnnotationURIHostPrefix:
+				m.host = value
+			case OAuthRedirectModelAnnotationResourceKindPrefix:
+				m.kind = value
+			case OAuthRedirectModelAnnotationResourceNamePrefix:
+				m.name = value
+			case OAuthRedirectModelAnnotationResourceGroupPrefix:
+				m.group = value
+			}
+			models[name] = m
+		}
+	}
+	return models
+}
+
+func parseModelPrefixName(key string) (string, string, bool) {
+	for _, prefix := range modelPrefixes {
+		if strings.HasPrefix(key, prefix) {
+			return prefix, key[len(prefix):], true
+		}
+	}
+	return "", "", false
+}
+
+func (a *saOAuthClientAdapter) extractRedirectURIs(modelsMap map[string]model, namespace string) redirectURIList {
+	var data redirectURIList
+	groupKindModelListMapper := map[unversioned.GroupKind]modelList{}
+	groupKindModelToURI := map[unversioned.GroupKind]namesToObjMapperFunc{
+		routeGroupKind: a.redirectURIsFromRoutes,
+	}
+
+	for _, model := range modelsMap {
+		gk := model.getGroupKind()
+		if _, ok := groupKindModelToURI[gk]; ok {
+			groupKindModelListMapper[gk] = append(groupKindModelListMapper[gk], model)
+		}
+	}
+
+	for gk, models := range groupKindModelListMapper {
+		if len(models) > 0 {
+			if names := models.getNames(); names.Len() > 0 {
```

Comment:
They don't check the same thing; it is possible for `len(models) > 0` and `names.Len() == 0`.